---
title: "Untitled"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F)
```

# calling resquired functions
```{r, message=F}
ensembl = biomaRt::useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", version = "100")
table_ensembl = biomaRt::getBM(attributes = c("ensembl_gene_id", "hgnc_symbol", "entrezgene_id", "uniprotswissprot", "ensembl_peptide_id"), mart = ensembl)

source("~/Dropbox/Work_Netherlands/FGFR3/R/Functions_FM_FGFR3_RE_annotation.R")
```

# libraries
```{r}
library(dplyr)
library(ggplot2)
library(gridExtra)
library(patchwork)
```

# data loading for FGFR3
Inframe fusion is defined by  
1) upstream and downstream exons are in-frame  
2) locations of the both breakends are in intron
```{r}
onco_data_FGFR3 = data.frame(R.utils::loadToEnv("~/Dropbox/Work_Netherlands/FGFR3/R/FM_FGFR3_RE_annotation.RData")[["onco_data"]])

FGFR3_trunc_REs = onco_data_FGFR3 %>% filter(FGFR3_RE_loc == "upstream")

FGFR3_trunc_REs_with_partners = FGFR3_trunc_REs %>% 
    filter(gene2_RE != "", !grepl(";", gene2_RE)) %>%
    mutate(Index = 1:nrow(.),
           gene2_EntrezID = table_ensembl$entrezgene_id[match(gene2_RE, table_ensembl$hgnc_symbol)],
           gene2_Ensembl = table_ensembl$ensembl_gene_id[match(gene2_RE, table_ensembl$hgnc_symbol)])

# genes unmatched symbol between bioMart and FM dataset
unmap_genes = c("AZI1" = "CEP131", "WHSC1" = "NSD2", "LARGE" = "LARGE1", "FAM129B" = "NIBAN2")

ind = match(FGFR3_trunc_REs_with_partners$gene2_RE, names(unmap_genes))
ind_1 = ind[!is.na(ind)]
ind_2 = which(!is.na(ind))

FGFR3_trunc_REs_with_partners$gene2_RE[ind_2] = unmap_genes[ind_1]
FGFR3_trunc_REs_with_partners = FGFR3_trunc_REs_with_partners %>% 
  mutate(gene2_EntrezID = table_ensembl$entrezgene_id[match(gene2_RE, table_ensembl$hgnc_symbol)],
         gene2_Ensembl = table_ensembl$ensembl_gene_id[match(gene2_RE, table_ensembl$hgnc_symbol)]) %>%
  mutate(gene2_EntrezID = ifelse(gene2_RE == "ZNF876P", 642280, gene2_EntrezID),
         gene2_Ensembl = ifelse(gene2_RE == "ZNF876P", "ENSG00000198155", gene2_Ensembl))

```

# data loading for FGFR2
```{r}
onco_data_FGFR2 = R.utils::loadToEnv("~/Dropbox/Work_Netherlands/FGFR/FGFR2_C_trunc/Nature_figures/data/FM/FM_FGFR2_vars.RData")[["onco_data_new"]]
onco_data_FGFR2 = data.frame(Index = 1:nrow(onco_data_FGFR2), onco_data_FGFR2)

# E18-truncated FGFR2 (including intergenic RE)
FGFR2_trunc_REs = onco_data_FGFR2 %>% 
    filter(!is.na(RE_FGFR2), RE_location == "FGFR2 is upstream", grepl("I17|E18", FGFR2_brkpt), !is.na(RE_pos1), !is.na(RE_pos2)) %>%
    mutate(gene1 = stringr::str_split_fixed(RE_partner, "-", 2)[,1],
           gene2 = stringr::str_split_fixed(RE_partner, "-", 2)[,2])

# there is inconsistent symbol between FMI and hgnc_symbol
unmap_genes = c("C10orf118" = "CCDC186", "C10orf76" = "ARMH3", "C7orf73" = "STMP1", "CCDC147" = "CFAP58", "KIAA1009" = "CEP162", "KIAA1524" = "CIP2A", "KIAA1598" = "SHTN1",
  "KIAA1967" = "CCAR2", "LINC00959" = "C10orf143", "MKL2" = "MRTFB", "MST4" = "STK26", "NARS" = "NARS1", "SEPT10" = "SEPTIN10", "WARS" = "WARS1", "WDR65" = "CFAP57",
"FAM213A" = "PRXL2A", "FAM194B" = "ERICH6B", "METTL10" = "EEF1AKMT2", "TUBGCP2;ZNF511" = "TUBGCP2", "LOC387723" = "C10orf143", "AZI1" = "AZIN1", "C11orf30" = "EMSY",
"LOC100132146" = "FAM240A")

# replace FMI gene name with HGNC symbol
ind = match(FGFR2_trunc_REs$gene2, names(unmap_genes))
ind_1 = ind[!is.na(ind)]
ind_2 = which(!is.na(ind))
FGFR2_trunc_REs$gene2[ind_2] = unmap_genes[ind_1]
FGFR2_trunc_REs = FGFR2_trunc_REs %>% 
  mutate(gene2_EntrezID = table_ensembl$entrezgene_id[match(gene2, table_ensembl$hgnc_symbol)],
         gene2_Ensembl = table_ensembl$ensembl_gene_id[match(gene2, table_ensembl$hgnc_symbol)])

FGFR2_trunc_REs[which(is.na(FGFR2_trunc_REs$gene2_EntrezID)),]
FGFR2_trunc_REs = FGFR2_trunc_REs %>% 
  mutate(gene2_EntrezID = replace(gene2_EntrezID, gene2 == "LINC00261", 140828), 
         gene2_EntrezID = replace(gene2_EntrezID, gene2 == "FAM24B-CUZD1", 100533195))

# E18-truncated FGFR2 (excluding intergenic RE)
FGFR2_trunc_REs_with_partners = FGFR2_trunc_REs %>%
    filter(RE_FGFR2 != "intergenic RE") %>%
    left_join(onco_data_FGFR2[, c(1:3)], by = "Index") %>%
    mutate(gene2_EntrezID = table_ensembl$entrezgene_id[match(gene2, table_ensembl$hgnc_symbol)],
           gene2_Ensembl = table_ensembl$ensembl_gene_id[match(gene2, table_ensembl$hgnc_symbol)])

partner_genes_FGFR2 = unique(FGFR2_trunc_REs_with_partners$gene2_EntrezID)
```


# oncoplot table
```{r}
#df_oncoplot = onco_data %>% filter(!grepl("intron 16", FGFR3_brkpt_loc))
df_oncoplot = onco_data_FGFR3

# FGFR3 mutation
FGFR3_mut = rep(NA, nrow(df_oncoplot))
FGFR3_mut[df_oncoplot$Type == "Mut_trunc"] = "E18-trunc"
FGFR3_mut[df_oncoplot$Type == "Mut_splice site"] = "E18-splice"

# FGFR3 amplifications
FGFR3_amp = readxl::read_excel("~/Dropbox/Work_Netherlands/FGFR3/FGFR3 Data Update 01022024.xlsx", sheet = 1)
FGFR3_amp = ifelse(df_oncoplot$deidentifiedSpecimenName %in% FGFR3_amp$deidentifiedSpecimenName, "Amp", NA)

# FGFR3 partial amplifications
t = readxl::read_excel("~/Dropbox/Work_Netherlands/FGFR3/FGFR3 Truncs.xlsx", sheet = 2)
t = df_oncoplot$deidentifiedSpecimenName %in% t$deidentifiedSpecimenName
FGFR3_amp[t == T] = "Partial amp"


# FGFR3 rearrangement
FGFR3_re_loc = rep(NA, nrow(df_oncoplot))
FGFR3_re_loc[grepl("chr4", df_oncoplot$pos1_RE) & grepl("chr4", df_oncoplot$pos2_RE)] = "Intrachromosomal"
FGFR3_re_loc[grepl("chr4", df_oncoplot$pos1_RE) & !grepl("chr4", df_oncoplot$pos2_RE)] = "Interchromosomal"
FGFR3_re_loc[!grepl("chr4", df_oncoplot$pos1_RE) & grepl("chr4", df_oncoplot$pos2_RE)] = "Interchromosomal"

# FGFR3 RE type
FGFR3_RE_in_strand = df_oncoplot %>% filter(grepl("Fusion", Type)) %>%
  mutate(effect = ifelse(grepl("inframe", breakpoint_5p_by_Jessica_annot) & grepl("inframe", breakpoint_3p_by_Jessica_annot),
                         "inframe", "in-strand (frame-unknown)"))

df_oncoplot = df_oncoplot %>% 
  mutate(RE.type = rep(NA, nrow(df_oncoplot)),
         RE.type = replace(RE.type, grepl("Fusion", Type), FGFR3_RE_in_strand$effect),
         RE.type = replace(RE.type, grepl("Out-of-strand RE", Type), "out-of-strand RE"),
         RE.type = replace(RE.type, grepl("RE with intergenic space", Type), "intergenic RE"),
         RE.type = replace(RE.type, grepl("Internal RE", Type), "re_internal"))

df_oncoplot = data.frame(Order = 1:nrow(df_oncoplot),
                         TCGA_type = df_oncoplot$TCGA,
                         FGFR3_mut = FGFR3_mut,
                         FGFR3_amp = FGFR3_amp,
                         RE_FGFR3 = df_oncoplot$RE.type,
                         RE_gene1 = df_oncoplot$gene1_RE,
                         RE_gene2 = df_oncoplot$gene2_RE,
                         FGFR3_brkpt = df_oncoplot$FGFR3_brkpt_loc,
                         FGFR3_RE_type = df_oncoplot$FGFR3_RE_loc,
                         FGFR3_chr_SV = FGFR3_re_loc,
                         RE_with_TACC3 = df_oncoplot$RE_with_TACC3,
                         TACC3_brkpt = df_oncoplot$TACC3_brkpt_loc_sum)

df_oncoplot$RE_FGFR3 = factor(df_oncoplot$RE_FGFR3, levels = c("inframe", "in-strand (frame-unknown)", "intergenic RE", "out-of-strand RE", "re_internal"))
df_oncoplot$FGFR3_brkpt = factor(df_oncoplot$FGFR3_brkpt, levels = c("intron 16", "exon 17", "intron 17", "exon 18", "exon 18_UTR"))
df_oncoplot$FGFR3_RE_type = factor(df_oncoplot$FGFR3_RE_type, levels = c("upstream", "downstream"))
df_oncoplot$FGFR3_chr_SV = factor(df_oncoplot$FGFR3_chr_SV, levels = c("Intrachromosomal", "Interchromosomal"))
df_oncoplot$TACC3_brkpt = factor(df_oncoplot$TACC3_brkpt, levels = c("intron 7", "intron 9", "intron 10", "others"))

# color code for TCGA
col_TCGA = c("BLCA" = "#000000", "Brain, GBM" = "#FFFF00", "LUSC" = "#b26120", "LUAD" = "#B3B3B3", "BRCA" = "#FF69B4", 
             "Unknown primary" = "#606060", "KIRC" = "#CCCCFF", "CESC" = "#00FF00", "HNSC" = "#E6AB02", "Brain, LGG" = "#FFFFCC", 
             "Brain, Astrocytoma" = "#D2EF92", "UCS" = "#990000", "CHOL" = "#377EB8", "COAD" = "#FDC086", 
             "STAD" = "#7FC97F", "PAAD" = "#556b2f", "OV" = "#F4CAE4", "ESCA" = "#ff4500", "READ" = "#0000ff",
             "SARC" = "#20b2aa", "Other" = "#2f4f4f")

df_oncoplot$TCGA_type = factor(df_oncoplot$TCGA_type, levels = names(col_TCGA))

ha = HeatmapAnnotation(df = df_oncoplot[,-c(1,6,7)], 
                       col = list(TCGA_type = col_TCGA, 
                                  FGFR3_mut = c("E18-trunc" = "#0C5D08", "E18-splice" = "#9BFF9D"),
                                  FGFR3_amp = c("Amp" = "black", "Partial amp" = "yellow"),
                                  RE_FGFR3 = c("inframe" = "#660000", "in-strand (frame-unknown)" = "#F90708", 
                                               "intergenic RE" = "#FF66FF", "out-of-strand RE" = "#FFD700", "re_internal" = "#989CFD"),
                                  FGFR3_brkpt = c("intron 16" = "#4469F2", "exon 17" = "#E8A8EF", "intron 17" = "#911eb4", "exon 18" = "#BFEF45", "exon 18_UTR" = "#567707"),
                                  FGFR3_RE_type = c("upstream" = "#A30059", "downstream" = "#006FA6"),
                                  FGFR3_chr_SV = c("Intrachromosomal" = "#469990", "Interchromosomal" = "#e6beff"),
                                  RE_with_TACC3 = c("TRUE" = "black", "FALSE" = "white"),
                                  TACC3_brkpt = c("intron 7" = "#00CC00", "intron 9" = "#0066CC", "intron 10" = "#990000", "others" = "#404040")), 
                       na_col = "white", annotation_legend_param = list(TCGA_type = list(nrow = 5)))
```

# Heatmap
```{r, fig.width=12, fig.height=4}

df_sort_v2= df_oncoplot %>% arrange(RE_FGFR3, FGFR3_RE_type, FGFR3_brkpt, FGFR3_chr_SV, FGFR3_mut)
df_1 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "inframe"),]
df_1 = df_1 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_1$TCGA_type), decreasing = T))))
df_2 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "in-strand (frame-unknown)"),]
df_2 = df_2 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_2$TCGA_type), decreasing = T))))
df_3 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "intergenic RE"),]
df_3 = df_3 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_3$TCGA_type), decreasing = T))))
df_4 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "out-of-strand RE"),]
df_4 = df_4 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_4$TCGA_type), decreasing = T))))
df_5 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "re_internal"),]
df_5 = df_5 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_5$TCGA_type), decreasing = T))))

df_RE = c(df_1$Order, df_2$Order, df_3$Order, df_4$Order, df_5$Order)

df_1 = df_sort_v2[which(df_sort_v2$FGFR3_mut == "E18-trunc" & !(df_sort_v2$Order %in% c(df_RE))),]
df_1 = df_1 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_1$TCGA_type), decreasing = T))))

df_2 = df_sort_v2[which(df_sort_v2$FGFR3_mut == "E18-splice" & !(df_sort_v2$Order %in% c(df_RE))),]
df_2 = df_2 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_2$TCGA_type), decreasing = T))))

df_Mut = c(df_1$Order, df_2$Order)

ind_ord = c(df_RE, df_Mut)

hm = Heatmap(matrix = matrix(nrow=0, ncol = length(ind_ord)), top_annotation = ha[ind_ord])

#pdf("~/Dropbox/Work_Netherlands/FGFR3/R/figures/FM_oncoplot.pdf", width = 12, height = 6)
draw(hm, annotation_legend_side = "bottom", merge_legend = F)
#dev.off()
```

# export information on total samples
```{r}
df_sort_v2$TACC3_brkpt_loc = onco_data_FGFR3$TACC3_brkpt_loc[match(df_sort_v2$Order, onco_data_FGFR3$Index)]
write.table(df_sort_v2, "~/Dropbox/Work_Netherlands/FGFR3/Manuscripts/Table_S_total_FMI_samples.txt", sep = "\t", row.names = F, col.names = T)
```

# Self-interacting domains
```{r}
uniprot_prot = read.table("~/Dropbox/Work_Netherlands/FGFR3/R/data/uniprot-filtered-organism__Homo+sapiens+(Human)+[9606]_+AND+review--.tab", 
                          header = T, sep = "\t", stringsAsFactors = F, fill = T, comment.char = "", quote = "") 

# SLIPPER SIP golden list
SLIPPER_SIP = readxl::read_xlsx("~/Dropbox/Work_Netherlands/FGFR3/R/data/Human SIP-direct_interaction-v20121201.xlsx", skip = 1, sheet = 1, col_names = T) %>%
  mutate(Interactor_A_entrez = table_ensembl$entrezgene_id[match(`Interactor_A （SwissProt AC）`, table_ensembl$uniprotswissprot)],
         Interactor_B_entrez = table_ensembl$entrezgene_id[match(`Interactor_B (SwissProt AC)`, table_ensembl$uniprotswissprot)])

# PPIDM database (domain-domain interactions)
PPIDM = read.csv2("~/Dropbox/Work_Netherlands/FGFR3/R/data//PPIDM_FullSortedDataset_84K_GSB.csv", sep = ",", stringsAsFactors = F, header = T)
PPIDM_self = PPIDM %>% filter(D1 == D2, CLASS == "Gold", IN_GOLDSTANDARD == "yes") %>% distinct()

# 3DID interacting domain
threedid = read.table("~/Dropbox/Work_Netherlands/FGFR3/R/data/3did_flat", header = F, sep = ",", stringsAsFactors = F, fill = T, comment.char = "", quote = "")
threedid = data.frame(stringr::str_split_fixed(threedid[grepl("#=ID", threedid[,1]),1], "\t", 5)[, 2:5]) %>%
  rename_all(~c("domain_1_name", "domain_2_name", "domain_1_pfam", "domain_2_pfam")) %>%
  mutate(domain_1_pfam = stringr::str_split_fixed(gsub("\\@Pfam| \\(", "", domain_1_pfam), "[.]", 2)[,1],
         domain_2_pfam = stringr::str_split_fixed(gsub("\\@Pfam|\\) ", "", domain_2_pfam), "[.]", 2)[,1])
threedid_self = unique(threedid %>% filter(domain_1_pfam == domain_2_pfam)) %>%
  distinct()

SID_ovp = intersect(PPIDM_self$D1, threedid_self$domain_1_pfam)
SID_ovp = data.frame(PfamID = SID_ovp, name = threedid_self$domain_1_name[match(SID_ovp, threedid_self$domain_1_pfam)])

uniprot_prot$SID_name = unlist(lapply(strsplit(uniprot_prot$Cross.reference..Pfam., ";"), function(x){
  paste0(SID_ovp$name[which(SID_ovp[,1] %in% x)], collapse = ";")
}))

#uniprot_prot = uniprot_prot %>% mutate(SID) %>% 
uniprot_prot = uniprot_prot %>%
  mutate(SIP_SLIPPER = Entry %in% SLIPPER_SIP$`Interactor_A （SwissProt AC）`,
         EntrezID = table_ensembl$entrezgene_id[match(Entry, table_ensembl$uniprotswissprot)],
         FGFR3_partner = EntrezID %in% unique(FGFR3_trunc_REs_with_partners$gene2_EntrezID),
         FGFR2_partner = EntrezID %in% unique(FGFR2_trunc_REs_with_partners$gene2_EntrezID)) %>%
  filter(!is.na(EntrezID))
```

# Self-interacting protein percentage
```{r, fig.width=10, fig.height=5}
library(ggplot2)
library(patchwork)

######################################################
# FGFR3
######################################################
# Proportion of SI proteins among FGFR3 known partners and non-partners
df_f3 = uniprot_prot %>% group_by(FGFR3_partner, SIP_SLIPPER) %>% summarise(n = n())

df_f3 = rbind(data.frame(df_f3), 
            data.frame(FGFR3_partner = "whole proteome", df_f3 %>% group_by(SIP_SLIPPER) %>% summarise(n = sum(n)))) %>%
  filter(FGFR3_partner != FALSE)

df_f3 = data.frame("FGFR3_partner" = c("partner", "background"),
                 "n" = (df_f3 %>% group_by(FGFR3_partner) %>% summarise(n = sum(n)) %>% pull(n)),
                 "x" = (df_f3 %>% filter(SIP_SLIPPER == TRUE) %>% pull(n)),
                 "SIP_perc" = (df_f3 %>% filter(SIP_SLIPPER == TRUE) %>% pull(n))*100 / (df_f3 %>% group_by(FGFR3_partner) %>% summarise(n = sum(n)) %>% pull(n)))

######################################################
# FGFR2
######################################################
df_f2 = uniprot_prot %>% group_by(FGFR2_partner, SIP_SLIPPER) %>% summarise(n = n())

df_f2 = rbind(data.frame(df_f2), 
            data.frame(FGFR2_partner = "whole proteome", df_f2 %>% group_by(SIP_SLIPPER) %>% summarise(n = sum(n)))) %>%
  filter(FGFR2_partner != FALSE)

df_f2 = data.frame("FGFR2_partner" = c("partner", "background"),
                 "n" = (df_f2 %>% group_by(FGFR2_partner) %>% summarise(n = sum(n)) %>% pull(n)),
                 "x" = (df_f2 %>% filter(SIP_SLIPPER == TRUE) %>% pull(n)),
                 "SIP_perc" = (df_f2 %>% filter(SIP_SLIPPER == TRUE) %>% pull(n))*100 / (df_f2 %>% group_by(FGFR2_partner) %>% summarise(n = sum(n)) %>% pull(n)))

######################################################
# plot SIP percentage - SLIPPER
######################################################
df_f2_f3 = data.frame(Type = c("whole proteome", "FGFR2 RE partners", "FGFR3 RE partners"),
                      rbind(df_f2[2:1,2:4], df_f3[1,2:4])) %>%
  mutate(Type = factor(Type, levels = c("whole proteome", "FGFR2 RE partners", "FGFR3 RE partners")))

# Percentage of self-interacting proteins (based on SLIPPER database) in background genes and FGFR3 fusion partners
p_bar_SIP = ggplot(df_f2_f3, aes(x = Type, y = SIP_perc)) + geom_bar(stat = "identity") + theme_classic() +
  theme(axis.text.x = element_text(color = "black", size = 12),
        axis.text.y = element_text(color = "black", size = 12)) +
  ggtitle("% of Self-interacting proteins (SLIPPER)") + ylim(c(0, 30))

######################################################
# plot intergenic RE percentage
######################################################
df_intergenic = data.frame(type = c("FGFR2", "FGFR3"),
                           x = c(length(which(FGFR2_trunc_REs$RE_FGFR2 == "intergenic RE")),
                                 sum(is.na(FGFR3_trunc_REs$gene2_RE))),
                           n = c(nrow(FGFR2_trunc_REs), nrow(FGFR3_trunc_REs))) %>%
  mutate(perc = 100*x/n)
p_bar_intergenic = ggplot(df_intergenic, aes(x = type, y = perc)) + geom_bar(stat = "identity") + theme_classic() +
  theme(axis.text.x = element_text(color = "black", size = 12),
        axis.text.y = element_text(color = "black", size = 12)) +
  ggtitle("% of intergenic RE") + ylim(c(0, 30))

p_bar_SIP + p_bar_intergenic


######################################################
# Proportion of Coiled-coil proteins (bassed on UniProt) among FGFR3 partners and background
######################################################
df_partner_f3 = uniprot_prot %>%
    filter(FGFR3_partner == T) %>% 
    mutate(CC_domain = ifelse(Coiled.coil != "", T, F)) %>%
    group_by(CC_domain) %>% summarise(n = n()) %>%
    mutate(freq = n / sum(n)) %>% filter(CC_domain == T)
df_background_f3 = uniprot_prot %>%
    mutate(CC_domain = ifelse(Coiled.coil != "", T, F)) %>%
    group_by(CC_domain) %>% summarise(n = n()) %>%
    mutate(freq = n / sum(n)) %>% filter(CC_domain == T)

df_f3 = rbind(data.frame(type = "background", df_background_f3),
           data.frame(type = "FGFR3 partners", df_partner_f3))

p_bar_CC_f3 = ggplot(df_f3, aes(x = type, y = freq)) + geom_bar(stat = "identity") + theme_classic() +
  theme(axis.text.x = element_text(color = "black", size = 12),
        axis.text.y = element_text(color = "black", size = 12)) +
  ggtitle("% of proteins with Coiled-coil")


# Proportion of proteins with coiled-coiled or SAM domains among FGFR3 partners and background
df_partner_f3 = uniprot_prot %>%
    filter(FGFR3_partner == T) %>% 
    mutate(CC_SAM_domain = ifelse(Coiled.coil != "" | grepl("PF00536", Cross.reference..Pfam.), T, F)) %>%
    group_by(CC_SAM_domain) %>% summarise(n = n()) %>%
    mutate(freq = n / sum(n)) %>% filter(CC_SAM_domain == T)
df_background_f3 = uniprot_prot %>%
    mutate(CC_SAM_domain = ifelse(Coiled.coil != "" | grepl("PF00536", Cross.reference..Pfam.), T, F)) %>%
    group_by(CC_SAM_domain) %>% summarise(n = n()) %>%
    mutate(freq = n / sum(n)) %>% filter(CC_SAM_domain == T)

df_f3 = rbind(data.frame(type = "background", df_background_f3),
           data.frame(type = "FGFR3 partners", df_partner_f3))

p_bar_CC_SAM_f3 = ggplot(df_f3, aes(x = type, y = freq)) + geom_bar(stat = "identity") + theme_classic() +
  theme(axis.text.x = element_text(color = "black", size = 12),
        axis.text.y = element_text(color = "black", size = 12)) +
  ggtitle("% of proteins with coiled-coil or SAM")


p_bar_SIP + p_bar_CC_f3 + p_bar_CC_SAM_f3
```

# FGFR2 partner distribution
```{r}
# pie chart for FGFR2 fusion partners
Var1 = FGFR2_trunc_REs %>% pull(gene2)
Var1[Var1=="N/A"] = "intergenic"
df = data.frame(table(Var1), stringsAsFactors = F) %>% arrange(desc(Freq))
df2 = rbind(df[df$Freq>9,], data.frame(Var1 = "others", Freq = sum(df$Freq[df$Freq<=9])))
df2$Partner = paste(df2$Var1, " (", df2$Freq, ")", sep = "")
df2$Partner = factor(df2$Partner, levels = rev(as.character(df2$Partner)))
p1 = ggplot(df2, aes(x = "", y = Freq, fill = Partner)) + geom_bar(stat = "identity", width = 1) +
  coord_polar("y", start = 0)  + theme_void() + guides(fill = guide_legend(reverse = TRUE))
p1

# FGFR2 partner in uniprot protein
uniprot_prot = uniprot_prot %>%
  mutate(FGFR2_partner = EntrezID %in% partner_genes_FGFR2)

# Proportion of Coiled-coil proteins (bassed on UniProt) among FGFR3 partners and background
df_partner_f2 = uniprot_prot %>%
    filter(FGFR2_partner == T) %>% 
    mutate(CC_domain = ifelse(Coiled.coil != "", T, F)) %>%
    group_by(CC_domain) %>% summarise(n = n()) %>%
    mutate(freq = n / sum(n)) %>% filter(CC_domain == T)

df_partner_f3 = uniprot_prot %>%
    filter(FGFR3_partner == T) %>% 
    mutate(CC_domain = ifelse(Coiled.coil != "", T, F)) %>%
    group_by(CC_domain) %>% summarise(n = n()) %>%
    mutate(freq = n / sum(n)) %>% filter(CC_domain == T)
df_background = uniprot_prot %>%
    mutate(CC_domain = ifelse(Coiled.coil != "", T, F)) %>%
    group_by(CC_domain) %>% summarise(n = n()) %>%
    mutate(freq = n / sum(n)) %>% filter(CC_domain == T)

df = rbind(data.frame(type = "background", df_background),
           data.frame(type = "FGFR2 partners", df_partner_f2),
           data.frame(type = "FGFR3 partners", df_partner_f3))

p_bar_CC_FGFR = ggplot(df, aes(x = type, y = freq)) + geom_bar(stat = "identity") + theme_classic() +
  theme(axis.text.x = element_text(color = "black", size = 12),
        axis.text.y = element_text(color = "black", size = 12)) +
  ggtitle("% of proteins with Coiled-coil_FGFR2")

#pdf("~/Dropbox/Work_Netherlands/FGFR3/R/figures/FM_FGFR3_partner_CC_domain.pdf", width = 7, height = 5)
p_bar_CC_FGFR
#dev.off()
```


# Distribution of FGFR3 partners
```{r}
donuts <- function(x, group = 1, labels = NA, col = NULL, radius = c(.7, 1)) {
  group <- rep_len(group, length(x))
  ug  <- unique(group)
  tbl <- table(group)[order(ug)]

  col <- if (is.null(col))
    seq_along(ug) else rep_len(col, length(ug))
  col.main <- Map(rep, col[seq_along(tbl)], tbl)
  col.sub  <- lapply(col.main, function(x) {
    al <- head(seq(0, 1, length.out = length(x) + 2L)[-1L], -1L)
    Vectorize(adjustcolor)(x, alpha.f = al)
  })

  plot.new()

  par(new = TRUE)
  pie(x, border = NA, radius = radius[2L],
      col = unlist(col.sub), labels = labels, cex = 0.7)

  par(new = TRUE)
  pie(x, border = NA, radius = radius[1L],
      col = unlist(col.main), labels = NA)
}
```

# Self-interacting domain (from PPIDM and 3did database) + uniprot coild-coil domain
```{r, fig.width=10, fig.height=10}
###############################################
# SID domain
###############################################

df = FGFR3_trunc_REs_with_partners %>%
    mutate(SID = gene2_EntrezID %in% (uniprot_prot %>% filter((SID_name != "") | Coiled.coil !="") %>% pull(EntrezID))) %>%
    mutate(group = ifelse(gene2_EntrezID %in% (uniprot_prot %>% filter(Coiled.coil !="") %>% pull(EntrezID)), "Coiled-coil", NA),
           group = replace(group, is.na(group) & SID== T, "SID"),
           group = replace(group, is.na(group), "Non-SID"),
           group = factor(group, levels = c("Coiled-coil", "SID", "Non-SID")))


df_1 = df %>%
    mutate(PARTNER_label = ifelse(gene2_RE %in% (names(which(table(df$gene2_RE)>=2))), gene2_RE, "others"),
           PARTNER_label = factor(PARTNER_label, levels = c(names(which(sort(table(df$gene2_RE), decreasing=T)>=2)), "others"))) %>%
    group_by(group, PARTNER_label) %>% summarise(n = n())

df_intergenic = FGFR3_trunc_REs %>% filter(is.na(gene2_RE)) %>%
    mutate(PARTNER_label = "intergenic", group = "intergenic") %>%
    group_by(group, PARTNER_label) %>% summarise(n = n())

df_comb = rbind(data.frame(df_1),
                data.frame(df_intergenic)) %>%
    mutate(group = factor(group, levels = c("Coiled-coil", "SID", "Non-SID", "intergenic")))

#pdf("~/FGFR3/nvim_R/figures/donuts_partners.pdf", width = 10, height = 10)
with(rev(df_comb), donuts(n, group, paste(PARTNER_label, " (", n, ")"), col = c("#F8766D", "#FFB266", "#C49A00", "#FF66FF", radius = c(0.3, 2))))
#dev.off()
```

# export domain information on FGFR3 partners
```{r}
fgfr3_partner_domains = df %>% group_by(group, gene2_RE) %>% summarise(n = n())
fgfr3_partner_domains$EntrezID = df$gene2_EntrezID[match(fgfr3_partner_domains$gene2_RE, df$gene2_RE)]
fgfr3_partner_domains = fgfr3_partner_domains %>% 
  mutate(uniprot_prot[match(EntrezID, uniprot_prot$EntrezID), c("Coiled.coil", "SID_name", "SIP_SLIPPER")])

fgfr3_partner_domains = rbind(fgfr3_partner_domains %>% filter(group == "Coiled-coil") %>% arrange(desc(n)),
                              fgfr3_partner_domains %>% filter(group == "SID") %>% arrange(desc(n)),
                              fgfr3_partner_domains %>% filter(group == "Non-SID") %>% arrange(desc(n)))

write.table(fgfr3_partner_domains, "~/Dropbox/Work_Netherlands/FGFR3/Manuscripts/Table_S2_FGFR3_partners_SID.txt", row.names = F, col.names = T, sep = "\t")
```


# enrichment of SV locations
```{r, fig.width=12, fig.height=4}
hg19_info = R.utils::loadToEnv("~/Dropbox/Work_Netherlands/FGFR3/R/HMF_data_purple_FGFR3.RData")[["hg19_info"]]
TXID = hg19_info$trx_gene_AA_major %>% filter(SYMBOL == "TACC3") %>% pull(TXID)

ind = as.numeric(TXID)
t_exon = hg19_info$trx_exon[[ind]]
t_intron = hg19_info$trx_intron[[ind]]
t_str = unique(as.character(strand(t_exon)))

if(t_str == "-"){
  t_intron = rev(t_intron)
}

df1 = data.frame(data.frame(t_exon)[, c(1,2,3,5)], id = paste("Exon", 1:length(t_exon)))
df2 = rbind(data.frame(data.frame(t_intron)[, c(1,2,3,5)], id = paste("Intron", 1:length(t_intron))), NA)
df_comb = (gdata::interleave(df1, df2))
df_comb = df_comb[-nrow(df_comb),]
df_comb$width = df_comb$end - df_comb$start

TACC3_SV_brkpt = onco_data_FGFR3 %>% filter(RE_with_TACC3 == "TRUE")

```

# breakpoint enrichment test - TACC3
```{r}
t_tab = table(TACC3_SV_brkpt$TACC3_brkpt_loc)

df_comb$n = as.numeric(t_tab[match(tolower(df_comb$id), names(t_tab))])
df_comb$n[is.na(df_comb$n)] = 0
df_comb$n_kb = (1000*df_comb$n)/(df_comb$width*nrow(TACC3_SV_brkpt))

# binomial test
prob_bino = df_comb$width/sum(df_comb$width)
pval_bino = numeric(nrow(df_comb))
for(i in 1:length(prob_bino)){
  t = binom.test(x = df_comb$n[i], n = sum(df_comb$n), p = prob_bino[i], alternative = "greater")
  pval_bino[i] = t$p.value
}

df_sv_bar = data.frame(lab = df_comb$id, n_kb = df_comb$n_kb, pval = -log10(pval_bino))
df_sv_bar_along = tidyr::gather(df_sv_bar, key="measure", value="value", c("n_kb", "pval"))
df_sv_bar_along$lab = factor(df_sv_bar_along$lab, levels = df_sv_bar$lab)

#pdf("~/Dropbox/Work_Netherlands/FGFR3/R/figures/binomial_test_TACC3.pdf", width = 10, height = 5)
ggplot(df_sv_bar_along, aes(x=lab, y=value)) + geom_bar(stat='identity')+
  facet_wrap(~measure,  ncol=1, scales = "free") + theme_bw() + 
  theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))
#dev.off()
```

# reannotate frame-unknown based on downstream partner location
```{r, fig.width=12, fig.height=4}
df_oncoplot = onco_data

# FGFR3 mutation
FGFR3_mut = rep(NA, nrow(df_oncoplot))
FGFR3_mut[df_oncoplot$Type == "Mut_trunc"] = "E18-trunc"
FGFR3_mut[df_oncoplot$Type == "Mut_splice site"] = "E18-splice"

# FGFR3 rearrangement
FGFR3_re_loc = rep(NA, nrow(df_oncoplot))
FGFR3_re_loc[grepl("chr4", df_oncoplot$pos1_RE) & grepl("chr4", df_oncoplot$pos2_RE)] = "Intrachromosomal"
FGFR3_re_loc[grepl("chr4", df_oncoplot$pos1_RE) & !grepl("chr4", df_oncoplot$pos2_RE)] = "Interchromosomal"
FGFR3_re_loc[!grepl("chr4", df_oncoplot$pos1_RE) & grepl("chr4", df_oncoplot$pos2_RE)] = "Interchromosomal"

# FGFR3 RE type
FGFR3_RE_in_strand = onco_data %>% filter(grepl("Fusion", Type)) %>%
  mutate(effect = ifelse(grepl("inframe", breakpoint_5p_by_Jessica_annot) & grepl("inframe", breakpoint_3p_by_Jessica_annot),
                         "inframe", "in-strand (frame-unknown)")) %>%
  mutate(effect = replace(effect, grepl("intron", breakpoint_3p_by_Jessica) & grepl("inframe", breakpoint_3p_by_Jessica_annot) & grepl("FGFR3", breakpoint_5p_by_Jessica), "inframe"))

df_oncoplot = df_oncoplot %>% 
  mutate(RE.type = rep(NA, nrow(df_oncoplot)),
         RE.type = replace(RE.type, grepl("Fusion", Type), FGFR3_RE_in_strand$effect),
         RE.type = replace(RE.type, grepl("Out-of-strand RE", Type), "out-of-strand RE"),
         RE.type = replace(RE.type, grepl("RE with intergenic space", Type), "intergenic RE"),
         RE.type = replace(RE.type, grepl("Internal RE", Type), "re_internal"))

df_oncoplot = data.frame(Order = 1:nrow(df_oncoplot),
                         TCGA_type = df_oncoplot$TCGA,
                         FGFR3_mut = FGFR3_mut,
                         RE_FGFR3 = df_oncoplot$RE.type,
                         FGFR3_brkpt = df_oncoplot$FGFR3_brkpt_loc,
                         FGFR3_RE_type = df_oncoplot$FGFR3_RE_loc,
                         FGFR3_chr_SV = FGFR3_re_loc,
                         RE_with_TACC3 = df_oncoplot$RE_with_TACC3,
                         TACC3_brkpt = df_oncoplot$TACC3_brkpt_loc_sum)

df_oncoplot$RE_FGFR3 = factor(df_oncoplot$RE_FGFR3, levels = c("inframe", "in-strand (frame-unknown)", "intergenic RE", "out-of-strand RE", "re_internal"))
df_oncoplot$FGFR3_brkpt = factor(df_oncoplot$FGFR3_brkpt, levels = c("intron 16", "exon 17", "intron 17", "exon 18", "exon 18_UTR"))
df_oncoplot$FGFR3_RE_type = factor(df_oncoplot$FGFR3_RE_type, levels = c("upstream", "downstream"))
df_oncoplot$FGFR3_chr_SV = factor(df_oncoplot$FGFR3_chr_SV, levels = c("Intrachromosomal", "Interchromosomal"))
df_oncoplot$TACC3_brkpt = factor(df_oncoplot$TACC3_brkpt, levels = c("intron 7", "intron 9", "intron 10", "others"))
df_oncoplot$TCGA_type = factor(df_oncoplot$TCGA_type, levels = names(col_TCGA))

ha = HeatmapAnnotation(df = df_oncoplot[,-1], 
                       col = list(TCGA_type = col_TCGA, 
                                  FGFR3_mut = c("E18-trunc" = "#0C5D08", "E18-splice" = "#9BFF9D"),
                                  FGFR3_amp = c("Amp" = "black", "Partial amp" = "yellow"),
                                  RE_FGFR3 = c("inframe" = "#660000", "in-strand (frame-unknown)" = "#F90708", 
                                               "intergenic RE" = "#FF66FF", "out-of-strand RE" = "#FFD700", "re_internal" = "#989CFD"),
                                  FGFR3_brkpt = c("intron 16" = "#4469F2", "exon 17" = "#E8A8EF", "intron 17" = "#911eb4", "exon 18" = "#BFEF45", "exon 18_UTR" = "#567707"),
                                  FGFR3_RE_type = c("upstream" = "#A30059", "downstream" = "#006FA6"),
                                  FGFR3_chr_SV = c("Intrachromosomal" = "#469990", "Interchromosomal" = "#e6beff"),
                                  RE_with_TACC3 = c("TRUE" = "black", "FALSE" = "white"),
                                  TACC3_brkpt = c("intron 7" = "#00CC00", "intron 9" = "#0066CC", "intron 10" = "#990000", "others" = "#404040")), 
                       na_col = "white", annotation_legend_param = list(TCGA_type = list(nrow = 5)))

df_sort_v2= df_oncoplot %>% arrange(RE_FGFR3, FGFR3_RE_type, FGFR3_brkpt, FGFR3_chr_SV, FGFR3_mut)

df_1 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "inframe"),]
df_1 = df_1 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_1$TCGA_type), decreasing = T))))
df_2 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "in-strand (frame-unknown)"),]
df_2 = df_2 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_2$TCGA_type), decreasing = T))))
df_3 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "intergenic RE"),]
df_3 = df_3 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_3$TCGA_type), decreasing = T))))
df_4 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "out-of-strand RE"),]
df_4 = df_4 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_4$TCGA_type), decreasing = T))))
df_5 = df_sort_v2[which(df_sort_v2$RE_FGFR3 == "re_internal"),]
df_5 = df_5 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_5$TCGA_type), decreasing = T))))

df_RE = c(df_1$Order, df_2$Order, df_3$Order, df_4$Order, df_5$Order)

df_1 = df_sort_v2[which(df_sort_v2$FGFR3_mut == "E18-trunc" & !(df_sort_v2$Order %in% c(df_RE))),]
df_1 = df_1 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_1$TCGA_type), decreasing = T))))

df_2 = df_sort_v2[which(df_sort_v2$FGFR3_mut == "E18-splice" & !(df_sort_v2$Order %in% c(df_RE))),]
df_2 = df_2 %>% arrange(factor(TCGA_type, levels = names(sort(table(df_2$TCGA_type), decreasing = T))))

df_Mut = c(df_1$Order, df_2$Order)

ind_ord = c(df_RE, df_Mut)

hm = Heatmap(matrix = matrix(nrow=0, ncol = length(ind_ord)), top_annotation = ha[ind_ord])
#pdf("~/FGFR/Daniel/R/files/FM_oncoplot.pdf", width = 12, height = 6)
draw(hm, annotation_legend_side = "bottom", merge_legend = F)
#dev.off()
```

```{r}
x = onco_data[match(df_sort_v2$Order, onco_data$Index), c("breakpoint_5p_by_Jessica", "breakpoint_3p_by_Jessica", "gene1_RE", "gene2_RE", g)]
export_table_FMI_trunc = cbind(df_sort_v2, x)
export_table_FMI_trunc = export_table_FMI_trunc %>%
  mutate(E18_trunc = if_else(!is.na(FGFR3_mut) | (FGFR3_RE_type == "upstream" & grepl("intron 17|exon 18", FGFR3_brkpt )), "TRUE", "FALSE"))


match(df_comb$PARTNER_label, uniprot_prot$Gene.names)
```

